yt/frontends/halo_catalog/io.py

```diff
@@ -104,9 +79,93 @@ def _count_particles(self, data_file):
+    """
+    Base class for io handlers to load halo member particles.
+    """
+
+    def _read_particle_coords(self, chunks, ptf):
+        pass
+
+    def _read_particle_fields(self, dobj, ptf):
+        # separate member particle fields from scalar fields
+        scalar_fields = defaultdict(list)
+        member_fields = defaultdict(list)
+        for ptype, field_list in sorted(ptf.items()):
+            for field in field_list:
+                if (ptype, field) in self.ds.scalar_field_list:
+                    scalar_fields[ptype].append(field)
+                else:
+                    member_fields[ptype].append(field)
+
+        all_data = self._read_scalar_fields(dobj, scalar_fields)
+        all_data.update(self._read_member_fields(dobj, member_fields))
+
+        for field, field_data in all_data.items():
+            yield field, field_data
+
+    # This will be refactored.
+    _read_particle_selection = IOHandlerGadgetFOFHaloHDF5._read_particle_selection
+
+
+class IOHandlerYTHalo(HaloDatasetIOHandler, IOHandlerYTHaloCatalog):
+    _dataset_type = "ythalo"
+
+    def _identify_fields(self, data_file):
+        with h5py.File(data_file.filename, "r") as f:
+            scalar_fields = [
+                ("halos", field) for field in f if not isinstance(f[field], h5py.Group)
+            ]
+            units = dict(
+                [(("halos", field), parse_h5_attr(f[field], "units")) for field in f]
+            )
+            if "particles" in f:
+                id_fields = [("halos", field) for field in f["particles"]]
+            else:
+                id_fields = []
+
+        return scalar_fields + id_fields, scalar_fields, id_fields, units
+
+    def _read_member_fields(self, dobj, member_fields):
+        all_data = defaultdict(lambda: np.empty(dobj.particle_number, dtype=np.float64))
+        if not member_fields:
+            return all_data
+        field_start = 0
+        for i, data_file in enumerate(dobj.field_data_files):
+            start_index = dobj.field_data_start[i]
+            end_index = dobj.field_data_end[i]
+            pcount = end_index - start_index
+            if pcount == 0:
+                continue
+            field_end = field_start + end_index - start_index
+            with h5py.File(data_file.filename, "r") as f:
+                for ptype, field_list in sorted(member_fields.items()):
+                    for field in field_list:
+                        field_data = all_data[(ptype, field)]
+                        my_data = f["particles"][field][start_index:end_index].astype(
+                            "float64"
+                        )
+                        field_data[field_start:field_end] = my_data
+            field_start = field_end
+        return all_data
+
+    def _read_scalar_fields(self, dobj, scalar_fields):
+        all_data = {}
+        if not scalar_fields:
+            return all_data
+        with h5py.File(dobj.scalar_data_file.filename, "r") as f:
```

Comment:
The h5py api has become less permissive
```suggestion
        with h5py.File(dobj.scalar_data_file.filename, mode="r") as f:
```